### PR TITLE
Allow host_connection_id in version metadata to be empty on ChanOpenInit

### PR DESCRIPTION
**Pull request by GNaD13**
## Description

closes: #5437

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **New Features**
	- Improved interchain accounts' registration and governance proposal execution process.
- **Refactor**
	- Enhanced code readability by refactoring version string setup and handshake logic.
- **Tests**
	- Updated and added test cases for ensuring consistency and correctness in version handling and metadata validation.
- **Documentation**
	- Clarified comments on host connection identifiers in metadata functions.

### CODE REVIEW

File: modules/apps/27-interchain-accounts/host/keeper/handshake_test.go
Comment (srdtrk) [inline modules/apps/27-interchain-accounts/host/keeper/handshake_test.go:56]: You shouldn't add an assert function since you can just check the `host_connection_id` after every successful handshake anyway.

File: modules/apps/27-interchain-accounts/types/metadata.go
Comment (srdtrk) [inline modules/apps/27-interchain-accounts/types/metadata.go:173]: This is not how we should do it. It is possible that we want to add other checks after this in the future. So, the only `nil` returned should be at the end of the function, not inside this if statement

**Review by srdtrk — CHANGES_REQUESTED** [review]
Thanks for the work @GNaD13. I noted some problems I had with the PR.

Also I'm wondering if we should change the `NewDefaultMetadata` function in controller's `OnChanInit` ([see here](https://github.com/cosmos/ibc-go/blob/677865d5a835bcf70baa33adde023fac32b92ef2/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L56)) to not require the `counterparty_connection_id`?

I could also take over the PR if you think these would be too much work for you @GNaD13 @crodriguezvega

**Comment by crodriguezvega** [comment]
> Also I'm wondering if we should change the `NewDefaultMetadata` function in controller's `OnChanInit` ([see here](https://github.com/cosmos/ibc-go/blob/677865d5a835bcf70baa33adde023fac32b92ef2/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L56)) to not require the `counterparty_connection_id`?

That sounds fine to me, but that change would be API breaking. Do you think it's ok for users to wait until the v9 to release this? Otherwise we could add a new function `NewDefaultMetadataWithoutHostConnection` (or something like that) only with the controller connection ID.

> I could also take over the PR if you think these would be too much work for you @GNaD13 @crodriguezvega

That's also fine by me. Is it ok for you if we continue from here, @GNaD13? Thank you very much for opening the PR!

File: modules/apps/27-interchain-accounts/types/metadata.go
Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/types/metadata.go:173]: I think we need to be a bit carefully with the validation of `metadata.HostConnectionId` because it will be valid to be empty in `ChanOpenInit` but it should not be empty anymore in `ChanOpenAck`, right?

**Comment by crodriguezvega** [comment]
> Yeah, that's also fine to me. It will be great if we could work together

Agree! Maybe then @srdtrk can comment the changes that would be needed and we can all together continue from there. :)

**Comment by srdtrk** [comment]
@crodriguezvega I think we can break API in this PR to main. We can worry about the API if we backport (we might not backport). So I'd say, we should break `NewDefaultMetadata`. Sounds good, then I'll continue to give suggestions under my review.

File: modules/apps/27-interchain-accounts/types/metadata.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/types/metadata.go:173]: I'd make the following suggestion @GNaD13 :
1. Revert your changes to validateConnectionParams.
2. In `ChanOpenTry`, before sending the metadata to validation, override `metadata.HostConnectionId` to be `connectionHops[0]` regardless of whether or not it is filled.
3. Send to validation after step 2.

Note that this is okay because the host is allowed to propose a new version string in `ChanOpenTry`. Does this sound good @crodriguezvega?

Comment (GNaD13) [reply modules/apps/27-interchain-accounts/types/metadata.go:173]: If so, I think we will also remove the validation in [ChanOpenInit](https://github.com/cosmos/ibc-go/blob/dc99ee43c1ade80094d56acd00ca16097b308d44/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L64)

**Comment by GNaD13** [comment]
> @crodriguezvega I think we can break API in this PR to main. We can worry about the API if we backport (we might not backport). So I'd say, we should break `NewDefaultMetadata`. Sounds good, then I'll continue to give suggestions under my review.

@srdtrk I've updated the code. Thanks for your suggestion. Could you take a look pls?

**Review by srdtrk — COMMENTED** [review]
Thanks for the improvements. I've added some followups. 

I'd also like you to update the end to end tests to use an empty `host_connection_id`. But I think we can open a separate PR for that if you don't want to.

File: modules/apps/27-interchain-accounts/host/keeper/handshake_test.go
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/host/keeper/handshake_test.go:1]: I added a check for this

File: modules/apps/27-interchain-accounts/types/metadata.go
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/types/metadata.go]: I removed it

**Comment by GNaD13** [comment]
> Thanks for the improvements. I've added some followups.
> 
> I'd also like you to update the end to end tests to use an empty `host_connection_id`. But I think we can open a separate PR for that if you don't want to.

@srdtrk I've updated the code. Could you take a look pls?

File: e2e/tests/interchain_accounts/utils.go
Comment (srdtrk) [inline e2e/tests/interchain_accounts/utils.go:7]: I'd like to pass an empty version string in the e2e test to ensure it works. So this function should not be needed and you can just use `NewDefaultMetadataString`

File: modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go
Comment (srdtrk) [inline modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go:25]: Can you move this to inside `SetupTest` since it seems it is the only place it is used?

**Review by srdtrk — COMMENTED** [review]
This looks very good. I added a few items. The only issue is that the callbacks module is not compiling since we broke the API. So you need to go into `modules/apps/callbacks` to fix it. After these changes, I'll give my approval.

**Comment by GNaD13** [comment]
> Please don't worry about the e2e tests failing related to channel upgrades

yeah, I got it

File: modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go]: I updated this code

File: e2e/tests/interchain_accounts/utils.go
Comment (srdtrk) [reply e2e/tests/interchain_accounts/utils.go]: I didn't mean you should change the function name. I meant that you could use the `NewDefaultMetadataString` function from icatypes instead of defining a new function here @GNaD13

File: modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/migrations/v6/migrations_test.go]: This is resolved

**Comment by srdtrk** [comment]
Please see https://github.com/cosmos/ibc-go/pull/5533#discussion_r1455253792

File: e2e/tests/interchain_accounts/utils.go
Comment (GNaD13) [reply e2e/tests/interchain_accounts/utils.go]: I got it

Comment (GNaD13) [reply e2e/tests/interchain_accounts/utils.go]: @srdtrk about the e2e test. If we use icatypes NewDefaultMetadataString with only controller port ID, it may get an error when ChanOpenInit check [here](https://github.com/decentrio/ibc-go/blob/ad3e34427557fc08f97caa01b727dc34a0d9317d/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L64). Should we use empty version so controller chain could create [metadata](https://github.com/decentrio/ibc-go/blob/ad3e34427557fc08f97caa01b727dc34a0d9317d/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L57) itself?

Comment (srdtrk) [reply e2e/tests/interchain_accounts/utils.go]: Oh, I see the issue now. Sorry about my oversight. We can go back to using your `DefaultTestMetadataVersionString`. I'll approve this PR after the e2e are passing

File: e2e/tests/upgrades/genesis_test.go
Comment (srdtrk) [inline e2e/tests/upgrades/genesis_test.go:23]: this doesn't cause a circular dependency right?

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go
Comment (srdtrk) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:50]: If connection is not used any longer, this should just be removed

**Review by srdtrk — APPROVED** [review]
Lgtm. Just added 1 more improvement request.

File: e2e/tests/upgrades/genesis_test.go
Comment (GNaD13) [reply e2e/tests/upgrades/genesis_test.go:23]: yeah, I've checked it. It doesn't cause a circular dependency

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go]: I removed these code

Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:50]: I removed it, since the connection is retrieved anyway in `ChanOpenInit`, so if it cannot be found, it will already error there.

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:39]: What is this test case actually testing?

Comment (GNaD13) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go]: This test is for my old code. I removed it. Thank you sir

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:45]: I am wondering now if we should just leave this as is. For our implementation of ICA controller it's easy to figure out the host connection ID, so we could just use it. Other implementations can leave it empty, if they cannot retrieve it, and with the change added to `OnChanOpenTry` in the host, then the host connection ID will be set. The change in `OnChanOpenTry` is actually what enables to leave the host connection ID empty in the proposed metadata of the controller. 

What do you think, @srdtrk?

File: modules/apps/27-interchain-accounts/types/metadata.go
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/types/metadata.go:37]: We could also just leave this function and `NewDefaultMetadataString` as they are, with two parameters, and if any implementation cannot provide a host connection ID, then they can just pass an empty string.

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:45]: Yeah, I think we could do that. Actually, that is my first approach. The only thing is we need to allow the [validation](https://github.com/cosmos/ibc-go/blob/97a52e35f35dadfc08aa9e852196e151eeaa56af/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L59) on controller chain that allow empty host connectionID

**Review by crodriguezvega — APPROVED** [review]
Thank you, @GNaD13. The changes look good to me now.

**Review by srdtrk — APPROVED** [review]

**Pull request merged**
